include/seqan3/range/view/pairwise_combine.hpp

```diff
@@ -0,0 +1,307 @@
+            return {*first_it, *second_it};
+        }
+        //!\}
+
+        /*!\name Arithmetic operators
+         * \{
+         */
+        //!\brief Pre-increment operator.
+        constexpr iterator_type & operator++(/*pre-increment*/)
+            noexcept(noexcept(++std::declval<source_iterator_type>()))
+        {
+            if (++second_it == end_it)
+            {
+                ++first_it;
+                second_it = first_it;
+                ++second_it;
+            }
+            return *this;
+        }
+
+        //!\brief Post-increment operator.
+        constexpr iterator_type operator++(int /*post-increment*/)
+            noexcept(noexcept(std::declval<source_iterator_type>()++))
+        {
+            iterator_type tmp{*this};
+            ++*this;
+            return tmp;
+        }
+        //!\}
+
+        /*!\name Comparison operators
+         * \{
+         */
+        friend constexpr bool operator==(iterator_type const & lhs, iterator_type const & rhs) noexcept
+        {
+            return std::tie(lhs.first_it, lhs.second_it) == std::tie(rhs.first_it, rhs.second_it);
+        }
+
+        friend constexpr bool operator!=(iterator_type const & lhs, iterator_type const & rhs) noexcept
+        {
+            return !(lhs == rhs);
+        }
+        //!\}
+
+    private:
+        //!\brief The iterator pointing to the first element of the pairwise combination.
+        source_iterator_type first_it{};
+        //!\brief The iterator pointing to the second element of the pairwise combination.
+        source_iterator_type second_it{};
+        //!\brief The end of the underlying range.
+        source_iterator_type end_it{};
+    };
+
+    //!\brief An alias for the class iterator type.
+    using iterator = iterator_type<std::ranges::iterator_t<source_range_type>>;
+
+public:
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr pairwise_combine_view() = default;
+    constexpr pairwise_combine_view(pairwise_combine_view const &) = default;
+    constexpr pairwise_combine_view(pairwise_combine_view &&) = default;
+    constexpr pairwise_combine_view & operator=(pairwise_combine_view const &) = default;
+    constexpr pairwise_combine_view & operator=(pairwise_combine_view &&) = default;
+    ~pairwise_combine_view() = default;
+
+    /*!\brief Constructs from a view.
+     * \param[in] range A rvalue reference to the underlying view.
+     *
+     * \details
+     *
+     * During construction the iterator pointing to the last element of the view is cached. This optimises
+     * the call to end if the underlying range models only ForwardRange. Otherwise the call to end will be
+     * linear in the number of elements of the underlying range.
+     */
+    template <typename other_range_type>
+        requires std::is_same_v<other_range_type, source_range_type>
+    explicit constexpr pairwise_combine_view(other_range_type && range) : src_range{std::move(range)}
```

Comment:
But I need to use the template argument. The issues with this is that there is some race with the move constructor but if it is not constrained  then anything can be put here. So I constrain it to be the same type as the source_range_type which is anyway only relevant if someone tries to call the constructor without deduction guides and out of its actual use with the view adaptor.

Reply:
ah ok. a comment would be good then :)
and you need to wrap the requires in `//!\cond`

Reply:
I kind of have to correct my previous statement. I thought I need the addtional template because of the forward reference, but in the case if I use the class template it is only an rvalue reference and this should be fine.